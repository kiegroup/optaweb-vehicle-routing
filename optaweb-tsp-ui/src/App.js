/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Component } from 'react';
import SockJS from 'sockjs-client';
import 'tachyons/css/tachyons.css';
import webstomp from 'webstomp-client';
import LocationList from './LocationList';
import TspMap from './TspMap';

class App extends Component {
  constructor() {
    super();

    this.state = {
      center: {
        lat: 49.23178,
        lng: 16.57561,
      },
      zoom: 5,
      counter: -1,
      selectedId: NaN,
      route: [],
      domicileId: NaN,
      distance: '',
      stompClient: null,
    };

    this.onClickLoad = this.onClickLoad.bind(this);
    this.onClickMap = this.onClickMap.bind(this);
    this.onClickRemove = this.onClickRemove.bind(this);
    this.onSelectLocation = this.onSelectLocation.bind(this);
  }

  componentDidMount() {
    this.connect('http://localhost:8080/tsp-websocket', () => {
      this.subscribe();
    });
  }

  onClickLoad() {
    this.state.stompClient.send('/app/demo');
  }

  onClickMap(e) {
    console.log(e.latlng);
    this.state.stompClient.send('/app/place', JSON.stringify(e.latlng));
  }

  onClickRemove(id) {
    if (id !== this.state.domicileId || this.state.route.length === 1) {
      this.state.stompClient.send(`/app/place/${id}/delete`);
    }
  }

  onSelectLocation(id) {
    this.setState({ selectedId: id });
  }

  connect(socketUrl, successCallback) {
    const webSocket = new SockJS(socketUrl);
    const stompClient = webstomp.over(webSocket, { debug: true });
    this.setState({ stompClient });
    stompClient.connect(
      {}, // no headers
      () => { // on connection, subscribe to the route topic
        successCallback();
      }, () => { // on error, schedule a reconnection attempt
        setTimeout(() => this.connect(socketUrl, successCallback), 1000);
      });
  }

  subscribe() {
    this.state.stompClient.subscribe('/topic/route', (message) => {
      const tsp = JSON.parse(message.body);
      this.setState({
        route: tsp.route,
        domicileId: tsp.route.length > 0 ? tsp.route[0].id : NaN,
        distance: tsp.distance,
      });
    });
  }

  render() {
    const { center, zoom, selectedId, route, domicileId, distance } = this.state;
    console.log(`Render, center: ${center}, route: [${route}], selected: ${selectedId}`);

    return (
      <div>
        <LocationList
          route={route}
          domicileId={domicileId}
          distance={distance}
          removeHandler={this.onClickRemove}
          selectHandler={this.onSelectLocation}
          loadHandler={this.onClickLoad}
        />
        <TspMap
          center={center}
          zoom={zoom}
          selectedId={selectedId}
          route={route}
          domicileId={domicileId}
          clickHandler={this.onClickMap}
          removeHandler={this.onClickRemove}
        />
      </div>
    );
  }
}

export default App;
