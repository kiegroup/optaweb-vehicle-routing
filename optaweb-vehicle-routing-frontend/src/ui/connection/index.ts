import ConnectedConnectionManager from './ConnectionManager';

export { ConnectedConnectionManager as ConnectionManager };
