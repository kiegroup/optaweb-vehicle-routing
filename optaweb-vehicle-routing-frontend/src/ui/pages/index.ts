import ConnectedDemo from './Demo';
import ConnectedRoute from './Route';
import ConnectedVehicles from './Vehicles';
import ConnectedVisits from './Visits';

export {
  ConnectedDemo as Demo,
  ConnectedVehicles as Vehicles,
  ConnectedVisits as Visits,
  ConnectedRoute as Route,
};
