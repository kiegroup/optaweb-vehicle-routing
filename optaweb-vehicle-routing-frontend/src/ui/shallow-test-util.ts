import { ReactElement } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createRenderer } from 'react-test-renderer/shallow';

export const shallow = (e: ReactElement): ReactElement => {
  const shallowRenderer = createRenderer();
  shallowRenderer.render(e);
  return shallowRenderer.getRenderOutput();
};

export const toJson = (e: ReactElement): ReactElement => e;
