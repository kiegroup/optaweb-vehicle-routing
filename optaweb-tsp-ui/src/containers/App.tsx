/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { UnpluggedIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import ConnectionError from 'src/components/ConnectionError';
import TravelingSalesmanProblem, {
  ITravelingSalesmanProblemProps,
} from '../components/TravelingSalesmanProblem';
import { IAppState } from '../store/configStore';
import { ITSPRouteWithSegments, tspOperations } from '../store/tsp/index';
import * as types from '../store/tsp/types';
import './App.css';

export interface IAppProps extends ITravelingSalesmanProblemProps {
  tsp: ITSPRouteWithSegments & types.IWSConnection;
  removeHandler: (id: number) => void;
  loadHandler: () => void;
  addHandler: (e: React.SyntheticEvent<HTMLElement>) => void;
  clearHandler: () => void;
}

const mapStateToProps = ({ tsp }: IAppState) => ({ tsp });

const mapDispatchToProps = (dispatch: Dispatch) => ({
  loadHandler() {
    dispatch(tspOperations.loadDemo());
  },
  addHandler(e: any) {
    dispatch(tspOperations.addLocation(e.latlng));
  },
  clearHandler() {
    dispatch(tspOperations.clearSolution());
  },
  removeHandler(id: number) {
    dispatch(tspOperations.deleteLocation(id));
  },
});

class App extends React.Component<IAppProps> {
  constructor(props: IAppProps) {
    super(props);
    this.onClickRemove = this.onClickRemove.bind(this);
  }

  onClickRemove(id: number) {
    const {
      tsp: { domicileId, route },
      removeHandler,
    } = this.props;
    if (id !== domicileId || route.length === 1) {
      removeHandler(id);
    }
  }

  render() {
    const { ws } = this.props.tsp;
    return (
      <div>
        {ws === types.WS_CONNECTION_STATE.ERROR && (
          <ConnectionError
            title="Oops... Connection error!"
            message="Please check your network connection."
            icon={<UnpluggedIcon />}
            help={
              'When connection is available the application will be functional again.'
            }
          />
        )}
        <TravelingSalesmanProblem {...this.props} />
      </div>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(App);
