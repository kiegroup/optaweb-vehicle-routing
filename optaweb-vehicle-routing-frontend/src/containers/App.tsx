/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { UnpluggedIcon } from '@patternfly/react-icons';
import * as React from 'react';
import { connect } from 'react-redux';
import ConnectionError from 'src/components/ConnectionError';
import TravelingSalesmanProblem from '../components/TravelingSalesmanProblem';
import { IAppState } from '../store/configStore';
import { websocketOperations } from '../store/websocket';
import { WebSocketConnectionStatus } from '../store/websocket/types';
import './App.css';

interface IStateProps {
  connectionStatus: WebSocketConnectionStatus;
}

interface IDispatchProps {
  connectClient: typeof websocketOperations.connectClient;
}

type Props = IStateProps & IDispatchProps;

const mapStateToProps = ({ connectionStatus }: IAppState): IStateProps => ({
  connectionStatus,
});

const mapDispatchToProps: IDispatchProps = {
  connectClient: websocketOperations.connectClient,
};

class App extends React.Component<Props> {
  constructor(props: Props) {
    super(props);
  }

  componentDidMount(): void {
    this.props.connectClient();
  }

  render() {
    const { connectionStatus } = this.props;
    return (
      <div>
        {connectionStatus === WebSocketConnectionStatus.ERROR && (
          <ConnectionError
            title="Oops... Connection error!"
            message="Please check your network connection."
            icon={<UnpluggedIcon />}
            help="When connection is available the application will be functional again."
          />
        )}
        <TravelingSalesmanProblem />
      </div>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
