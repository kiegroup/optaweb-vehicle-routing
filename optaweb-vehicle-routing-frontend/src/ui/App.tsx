/*
 * Copyright 2018 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Page, PageSection } from '@patternfly/react-core';
import * as React from 'react';
import { Route, Switch } from 'react-router-dom';
import Alerts from 'ui/components/Alerts';
import './App.css';
import Background from './components/Background';
import { ConnectionManager } from './connection';
import Header from './header/Header';
import { Demo, Route as RoutePage, Vehicles, Visits } from './pages';

export const pagesByPath = [
  { path: { canonical: '/demo', aliases: ['/'] }, page: Demo, label: 'Demo' },
  { path: { canonical: '/vehicles', aliases: [] }, page: Vehicles, label: 'Vehicles' },
  { path: { canonical: '/visits', aliases: [] }, page: Visits, label: 'Visits' },
  { path: { canonical: '/routes', aliases: [] }, page: RoutePage, label: 'Routes' },
];

const App: React.FC = () => (
  <>
    <ConnectionManager />
    <Alerts />
    <Page header={<Header />}>
      <Background />
      <PageSection
        style={{
          display: 'flex',
          flexDirection: 'column',
          overflowY: 'auto',
          height: '100%',
        }}
      >
        <Switch>
          {pagesByPath.map(({ path, page }) => ([path.canonical, ...path.aliases].map((p) => (
            <Route
              key={p}
              path={p}
              exact
              component={page}
            />
          ))))}
        </Switch>
      </PageSection>
    </Page>
  </>
);

export default App;
