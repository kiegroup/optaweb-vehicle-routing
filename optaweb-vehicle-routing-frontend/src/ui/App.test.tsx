import { shallow, toJson } from 'ui/shallow-test-util';
import App from './App';

describe('App', () => {
  it('should render correctly', () => {
    const app = shallow(<App />);
    expect(toJson(app)).toMatchSnapshot();
  });
});
