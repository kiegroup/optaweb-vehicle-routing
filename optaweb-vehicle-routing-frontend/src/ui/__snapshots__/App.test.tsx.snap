// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<Fragment>
  <Connect(ConnectionManager) />
  <Page
    breadcrumb={null}
    className=""
    header={<Header />}
    isManagedSidebar={false}
    onPageResize={null}
    sidebar={null}
    skipToContent={null}
  >
    <Background />
    <PageSection
      className=""
      noPadding={false}
      noPaddingMobile={false}
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "height": "100%",
          "overflowY": "auto",
        }
      }
      variant="default"
    >
      <Switch>
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(Vehicles)",
              "type": [Function],
            }
          }
          exact={true}
          path="/vehicles"
        />
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(Visits)",
              "type": [Function],
            }
          }
          exact={true}
          path="/visits"
        />
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(Route)",
              "type": [Function],
            }
          }
          exact={true}
          path="/route"
        />
        <Route
          component={
            Object {
              "$$typeof": Symbol(react.memo),
              "WrappedComponent": [Function],
              "compare": null,
              "displayName": "Connect(Demo)",
              "type": [Function],
            }
          }
          exact={true}
          path="/demo"
        />
      </Switch>
    </PageSection>
  </Page>
</Fragment>
`;
